File: python/121-130/Word Ladder.py
class Solution:
    # @param {string} beginWord
    # @param {string} endWord
    # @param {set<string>} wordDict
    # @return {integer}
    def ladderLength(self, beginWord, endWord, wordDict):
        dis = {}
        dis[beginWord] = 1
        chars = string.ascii_lowercase
        myqueue = [beginWord]
        
        while len(myqueue) > 0:
            word = myqueue[0]
            del myqueue[0]
            if word == endWord:
                break
            for i in range(len(word)):
                for c in chars:
                    newWord = word[:i] + c + word[i+1:]
                    if newWord in wordDict and newWord not in dis:
                        dis[newWord] = dis[word]+1
                        myqueue.append(newWord)
    
        if endWord not in dis:
            return 0
return dis[endWord]

File: cpp/121-130/Word Ladder.cpp

   
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> dict(begin(wordList), end(wordList));
        queue<string> q;
        q.push(beginWord);
        unordered_set<string> visited;
        visited.insert(beginWord);
        int step = 1;
        while (!q.empty()) {
            int k = q.size();
            for (int i = 0; i < k; i++) {
                string word = q.front();
                q.pop();
                if (word == endWord) {
                    return step;
                }
                for (int i = 0; i < word.size(); i++) {
                    char ch = word[i];
                    for (int j = 0; j < 26; j++) {
                        word[i] = 'a' + j;
                        if (visited.count(word) == 0 && dict.count(word) > 0) {
                            visited.insert(word);
                            q.push(word);
                        }
                    }
                    word[i] = ch;
                }
            }
            
            step++;
        }
        
        return 0;
    }
};
